presto-main/src/main/java/com/facebook/presto/operator/GroupedTopNBuilder.java

```diff
@@ -0,0 +1,654 @@
+                    peerGroupCounts.set(j, peerGroupCount);
+                }
+            }
+        };
+    }
+
+    public void compact()
+    {
+        for (int i = 0; i < currentPageCount; i++) {
+            if (pageReferences.get(i) != null) {
+                pageReferences.get(i).compact();
+            }
+        }
+        compactPages(
+                new AbstractIterator<Integer>()
+                {
+                    private int pageId;
+
+                    @Override
+                    protected Integer computeNext()
+                    {
+                        for (int i = pageId; i < currentPageCount; i++) {
+                            if (pageReferences.get(i) != null) {
+                                pageId = i + 1;
+                                return i;
+                            }
+                        }
+                        return endOfData();
+                    }
+                });
+    }
+
+    public long getEstimatedSizeInBytes()
+    {
+        long hashMemorySizeInBytes = groupByHash == null ? 0 : groupByHash.getEstimatedSize();
+        return INSTANCE_SIZE +
+                memorySizeInBytes +
+                hashMemorySizeInBytes +
+                groupedRows.sizeOf() +
+                pageReferences.sizeOf() +
+                emptyPageReferenceSlots.getEstimatedSizeInBytes();
+    }
+
+    @VisibleForTesting
+    List<Page> getBufferedPages()
+    {
+        ImmutableList.Builder<Page> builder = new ImmutableList.Builder<>();
+        for (int i = 0; i < currentPageCount; i++) {
+            if (pageReferences.get(i) != null) {
+                builder.add(pageReferences.get(i).getPage());
+            }
+        }
+        return builder.build();
+    }
+
+    @VisibleForTesting
+    void enablePeerGroupInfo()
+    {
+        computePeerGroupInfo = true;
+    }
+
+    private void processPage(Page newPage, @Nullable GroupByIdBlock groupIds)
+    {
+        verify(groupIds != null || groupByHash == null);
+
+        // save the new page
+        PageReference newPageReference = new PageReference(newPage);
+        memorySizeInBytes += newPageReference.getEstimatedSizeInBytes();
+        int newPageId;
+        if (emptyPageReferenceSlots.isEmpty()) {
+            // all the previous slots are full; create a new one
+            pageReferences.ensureCapacity(currentPageCount + 1);
+            newPageId = currentPageCount;
+            currentPageCount++;
+        }
+        else {
+            // reuse a previously removed page's slot
+            newPageId = emptyPageReferenceSlots.dequeueInt();
+        }
+        verify(pageReferences.get(newPageId) == null, "should not overwrite a non-empty slot");
+        pageReferences.set(newPageId, newPageReference);
+
+        // update the affected heaps and record candidate pages that need compaction
+        Set<Integer> pagesToCompact = new IntOpenHashSet();
```

Comment:
IIRC for the fastutil classes you want to use their interfaces.  Specifically for this class you want use `IntSet`, and then you can use their methods `add(int)` and `IntIterator`